### DISCUSSION TITLE
iped-hashdb.jar slow

### DISCUSSION THREAD

**Discussion by paulobreim** [sepinf-inc/IPED#2155] [Q&A]
Hi @wladimirleite ,

I downloaded the 4 files I found at NIST: RDS_2024.03.1_android, RDS_2024.03.1_ios, RDS_2024.03.1_legacy, RDS_2024.03.1_modern to create the hash base for IPED
Then I put the first one to run, Android, using the command:
jre\bin\java.exe -jar lib\iped-hashdb.jar -d k:\RDS_2024.03.1_android -o j:\iped-hashes\iped-hashes.db (all SSD)
But after 48 hours it looks like this:
Reading NSRL_DB file k:\RDS_2024.03.1_android\RDS_2024.03.1_android.db...
[========================64.0%======>

Assuming that what will be generated is a SQL database, which I haven't seen exactly what fields it has yet, I'm finding this very time-consuming.

tks
Paulo

**Comment by lfcnassif** [comment]
Hi @paulobreim. Just a guess, but try to disable your antivirus on both folders, adding an exception to them, it may help.

**Reply by wladimirleite** [reply]
I am running some tests here, and will reply later today.

**Reply by wladimirleite** [reply]
Hi @paulobreim!
I just processed "RDS_2024.03.1_android_minimal.db" and it took ~23 minutes here.
Same command you used (`java -jar lib\iped-hashdb.jar -d g:\RDS_2024.03.1_android_minimal.db -o d:\iped-hashes.db`), but used the "**minimal**" version as input.
```
Connected to database d:\iped-hashes.db
Database tables and indexes created.
Last HASH_ID = 0
Last PROPERTY_ID = 0
Properties loaded = 0

Reading NSRL_DB file g:\RDS_2024.03.1_android_minimal.db...
88530048 records read in 1373 seconds.
29406404 hashes inserted.
1 zero length hash was ignored.
59123643 records combined.

Commiting changes...
Commit completed in 11 seconds.
```

I always use these "minimal" versions as the "full" versions contain mostly redundant information (from the IPED's usage point of view), and they are much larger, take longer to download and import, larger generated databases etc.
I am downloading the latest Android full version (it will take a few hours), just to check how long would it take to be processed, and check if there is a way to speed up full version processing. Importing the "minimal" version is already taking a reasonable time (given the number of records, input and output size).
I strongly recommend that you try using the minimal version (at least to check how long it would take to be imported in your environment).
Check @lfcnassif recommendation (to disable antivirus), and I would also add: **disable Windows indexing** for the used folder and **do not use a volume where the SO is** (which seems you are already doing).

**Reply by paulobreim** [reply]
Hi @lfcnassif, Everything is off.
Without a doubt there must be some anomaly because since the first message I posted it has only moved 0.2%.
I'll do this test that @wladimirleite  did, with minimal, and see the results.
But I'm biased towards understanding this code to see the cause of the delay.

**Reply by lfcnassif** [reply]
@paulobreim you have a machine with 256GB or more RAM right? In this situation, creating a RAM drive and using it as the destination of the hash database while it is being built could help. After importing all hash sets, you can copy the final DB to your SSD and unmount the RAM drive.

**Reply by lfcnassif** [reply]
I would also look at the Resource Monitor for disk usage activities by unexpected processes.

**Reply by lfcnassif** [reply]
And also run a disk performance test on the destination SSD to check its heath.

**Reply by paulobreim** [reply]
@lfcnassif What software on Windows do you use to check the health of the SSD?

**Reply by lfcnassif** [reply]
Last time I used CrystalDiskInfo and Windows native Disk Management utility.

**Reply by lfcnassif** [reply]
@paulobreim did you see #2160?

**Reply by paulobreim** [reply]
I did a test with minimal (the same as wladimir) creating 2 HDs in RAM and it took 23 minutes.
Unless Wladimir has much faster equipment, I still found it slow.
CPU usage was just 4%
I will check #2160

**Reply by lfcnassif** [reply]
@wladimirleite's importing of Android minimal DB also took ~23min. It's totally acceptable to me, since it is an one time operation you just need to run once before processing your cases in this machine. You can also copy the resulting DB to other processing machines.

**Reply by wladimirleite** [reply]
@paulobreim, importing these hash sets is mostly I/O bound. I am using a standard SSD to store the output database (not the faster NVMe I use for the IPED temporary folder), and the input files are in a HDD.
There are 88,530,048 records in the Android minimal DB, so 23 minutes means ~64,000 records per second, which seems pretty fast for SQLite database operations.
